moveit_ros/moveit_servo/src/servo_node.cpp

```diff
@@ -53,111 +76,231 @@ ServoNode::ServoNode(const rclcpp::NodeOptions& options)
+  trajectory_publisher_ = node_->create_publisher<trajectory_msgs::msg::JointTrajectory>(
+      servo_params_.command_out_topic, rclcpp::SystemDefaultsQoS());
+
+  // Create status publisher
+  status_publisher_ =
+      node_->create_publisher<std_msgs::msg::Int8>(servo_params_.status_topic, rclcpp::SystemDefaultsQoS());
+
+  // Create service to enable switching command type
+  switch_command_type_ = node_->create_service<moveit_msgs::srv::ServoCommandType>(
+      "moveit_servo/switch_command_type",
+      std::bind(&ServoNode::switchCommandType, this, std::placeholders::_1, std::placeholders::_2));
+
+  // Create service to pause/unpause servoing
+  pause_servo_ = node_->create_service<std_srvs::srv::SetBool>("moveit_servo/pause_servo",
+                                                               std::bind(&ServoNode::pauseServo, this,
+                                                                         std::placeholders::_1, std::placeholders::_2));
+
+  // Start the servoing loop
+  servo_loop_thread_ = std::thread(&ServoNode::servoLoop, this);
 }
 
-void ServoNode::validateParams(const servo::Params& servo_params)
+void ServoNode::pauseServo(const std::shared_ptr<std_srvs::srv::SetBool::Request> request,
+                           const std::shared_ptr<std_srvs::srv::SetBool::Response> response)
 {
-  bool has_error = false;
-  if (servo_params.hard_stop_singularity_threshold <= servo_params.lower_singularity_threshold)
+  servo_paused_ = request->data;
+  response->success = (servo_paused_ == request->data);
+  if (servo_paused_)
   {
-    RCLCPP_ERROR(LOGGER, "Parameter 'hard_stop_singularity_threshold' "
-                         "should be greater than 'lower_singularity_threshold.' "
-                         "Check the parameters YAML file used to launch this node.");
-    has_error = true;
+    servo_->setCollisionChecking(false);
+    response->message = "Servoing disabled";
   }
-
-  if (!servo_params.publish_joint_positions && !servo_params.publish_joint_velocities &&
-      !servo_params.publish_joint_accelerations)
+  else
   {
-    RCLCPP_ERROR(LOGGER, "At least one of publish_joint_positions / "
-                         "publish_joint_velocities / "
-                         "publish_joint_accelerations must be true. "
-                         "Check the parameters YAML file used to launch this node.");
-    has_error = true;
+    servo_->setCollisionChecking(true);
+    response->message = "Servoing enabled";
   }
+}
 
-  if ((servo_params.command_out_type == "std_msgs/Float64MultiArray") && servo_params.publish_joint_positions &&
-      servo_params.publish_joint_velocities)
+void ServoNode::switchCommandType(const std::shared_ptr<moveit_msgs::srv::ServoCommandType::Request> request,
+                                  const std::shared_ptr<moveit_msgs::srv::ServoCommandType::Response> response)
+{
+  const bool is_valid = (request->command_type >= 0) && (request->command_type <= 2);
```

Comment:
Use the enum names instead of hard-coded numbers here.

Reply:
Done